perf_bench_test.go

```diff
@@ -24,127 +24,567 @@ package main
+	var (
+		sum         float64
+		funcs       = strings.Split(*funcNames, ",")
+		serviceTime = map[string]float64{
+			"helloworld":   1,
+			"chameleon":    12,
+			"pyaes":        1,
+			"image_rotate": 3,
+			"json_serdes":  6,
+			"lr_serving":   1,
+			"cnn_serving":  60,
+			"rnn_serving":  10,
+		}
+	)
+
+	for _, funcName := range funcs {
+		sum += serviceTime[funcName]
+	}
+
+	return sum / float64(len(funcs))
+}
+
+func calculateRPS(vmNum int) int {
+	baseRPS := getCPUIntenseRPS()
+	return vmNum * baseRPS
+}
+
+func TestBindSocket(t *testing.T) {
+	var (
+		servedTh      uint64
+		pinnedFuncNum int
+		isSyncOffload bool = true
+		testImage          = []string{"vhiveease/helloworld:var_workload"}
+	)
+
+	type testCase struct {
+		vmNum    int
+		expected string
+	}
+
+	cases := []testCase{
+		{vmNum: 1, expected: "13"},
+		{vmNum: 32, expected: "1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63"},
+	}
+
+	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
+	plrfncPid := pullImages(t, testImage)
+
+	*isBindSocket = true
+
+	for _, tCase := range cases {
+		testName := fmt.Sprintf("vmNum=%d", tCase.vmNum)
+		t.Run(testName, func(t *testing.T) {
+			bootVMs(t, testImage, 0, tCase.vmNum, plrfncPid)
+
+			pidBytes, err := getFirecrackerPid()
+			require.NoError(t, err, "Cannot get Firecracker PID")
+			vmPidList := strings.Split(string(pidBytes), " ")
+
+			for _, vm := range vmPidList {
+				vm = strings.TrimSpace(vm)
+				if vm != plrfncPid {
+					cpuBytes, err := exec.Command("taskset", "-cp", vm).Output()
+					require.NoError(t, err, "Cannot get VM PID")
+					cpuAffinity := strings.TrimSpace(strings.Split(string(cpuBytes), ":")[1])
+					require.Equal(t, tCase.expected, cpuAffinity, "VM was not binded to socket 1")
+				}
+			}
+
+			tearDownVMs(t, testImage, tCase.vmNum, isSyncOffload)
+		})
+	}
+}
+
+func bindSocket(omittedPid string) error {
+	var (
+		core1Free = true
+		cores     = 32
+	)
+	pidBytes, err := getFirecrackerPid()
+	if err != nil {
+		return err
+	}
+
+	vmPidList := strings.Split(string(pidBytes), " ")
+	for _, vm := range vmPidList {
+		vm = strings.TrimSpace(vm)
+		if vm != omittedPid {
+			if cores > len(vmPidList) {
+				if core1Free {
+					log.Debugf("binding pid: %s to core 1", vm)
+					cmd := exec.Command("taskset", "--all-tasks", "-cp", "13", vm)
+					if err := cmd.Run(); err != nil {
+						return err
+					}
+					core1Free = false
+				} else {
+					log.Debugf("binding pid: %s", vm)
+					cmd := exec.Command("taskset", "--all-tasks", "-cp", "1-11:2,15-63:2", vm)
+					if err := cmd.Run(); err != nil {
+						return err
+					}
+				}
+			} else {
+				log.Debugf("binding pid: %s", vm)
+				cmd := exec.Command("taskset", "--all-tasks", "-cp", "1-63:2", vm)
+				if err := cmd.Run(); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func checkInputValidation(t *testing.T) {
```

Comment:
rename: validateRuntimeArguments

you should write the comments in a form of "X is x but must be more/less/equal ...". Users would understand how they need to select the arguments.

Reply:
Done